src/libdredd/src/mutation_replace_expr.cc

```diff
@@ -0,0 +1,221 @@
+
+#include <cassert>
+#include <sstream>
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/LangOptions.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include "libdredd/util.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace dredd {
+dredd::MutationReplaceExpr::MutationReplaceExpr(const clang::Expr& expr)
+    : expr_(expr) {}
+
+std::string MutationReplaceExpr::GetFunctionName() const {
+  std::string result = "__dredd_replace_expr_";
+
+  // A string corresponding to the expression forms part of the name of the
+  // mutation function, to differentiate mutation functions for different
+  // types
+  if (expr_.getType()->isBooleanType()) {
+    result += "Bool";
+  } else if (expr_.getType()->isIntegerType()) {
+    result += "Int";
+  } else if (expr_.getType()->isFloatingType()) {
+    result += "Float";
+  } else {
+    assert(false && "Unsupported opcode");
+  }
+
+  if (expr_.isLValue()) {
+    result += "_lvalue";
+  }
+
+  return result;
+}
+
+std::string MutationReplaceExpr::GenerateMutatorFunction(
+    clang::ASTContext& ast_context, const std::string& function_name,
+    const std::string& result_type, int& mutation_id) const {
+  std::stringstream new_function;
+  new_function << "static " << result_type << " " << function_name << "(";
+  if (ast_context.getLangOpts().CPlusPlus) {
+    new_function << "std::function<" << result_type << "()>";
+  } else {
+    new_function << result_type;
+  }
+  new_function << " arg, int local_mutation_id) {\n";
+
+  std::string arg_evaluated = "arg";
+  if (ast_context.getLangOpts().CPlusPlus) {
+    arg_evaluated += "()";
+  }
+
+  // Quickly apply the original operator if no mutant is enabled (which will be
+  // the common case).
+  new_function << "  if (!__dredd_some_mutation_enabled) return "
+               << arg_evaluated << ";\n";
+
+  int mutant_offset = 0;
+  const clang::BuiltinType* exprType =
+      expr_.getType()->getAs<clang::BuiltinType>();
+
+  if (expr_.isLValue() && ast_context.getLangOpts().CPlusPlus) {
+    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                 << mutant_offset << ")) return ++" << arg_evaluated << ";\n";
+    mutant_offset++;
+
+    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                 << mutant_offset << ")) return --" << arg_evaluated << ";\n";
+    mutant_offset++;
+  }
+
+  if (!expr_.isLValue() &&
+      (exprType->isBooleanType() || exprType->isInteger())) {
+    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                 << mutant_offset << ")) return !(" << arg_evaluated << ");\n";
+    mutant_offset++;
+
+    new_function << "  if (__dredd_enabled_mutation(local_mutation_id + "
+                 << mutant_offset << ")) return ~(" << arg_evaluated << ");\n";
+    mutant_offset++;
+  }
```

Comment:
As per my comment on binary expression mutation, I think it would be good to extract this into a function.

And I suppose all the logic here, to decide when to insert a unary, should be the same as the logic in the binary and unary operator classes.

We could either do a refactoring in a follow-up PR, or you could make this a static method of  this class and have the other classes call it.

Reply:
Done as part of the refactoring described above.